examples/platform/esp32/mode-support/static-supported-modes-manager.cpp

```diff
@@ -0,0 +1,179 @@
+
+            memset(keyBuf, 0, sizeof(char) * ESP32Config::kMaxConfigKeyNameLength);
+            VerifyOrReturnValue(ESP32Config::KeyAllocator::SemanticTagMfgCode(keyBuf, sizeof(keyBuf), endpointId, index, stIndex) ==
+                                    CHIP_NO_ERROR,
+                                ModeOptionsProvider(nullptr, nullptr));
+            ESP32Config::Key stMfgCodeKey(ESP32Config::kConfigNamespace_ChipFactory, keyBuf);
+            VerifyOrReturnValue(ESP32Config::ReadConfigValue(stMfgCodeKey, semanticTagMfgCode) == CHIP_NO_ERROR,
+                                ModeOptionsProvider(nullptr, nullptr));
+
+            tag.value             = static_cast<uint16_t>(semanticTagValue);
+            tag.mfgCode           = static_cast<chip::VendorId>(semanticTagMfgCode);
+            semanticTags[stIndex] = static_cast<const SemanticTag>(tag);
+        }
+
+        option.label        = chip::CharSpan::fromCharString(modeLabelList[index].supportedModeLabel);
+        option.mode         = static_cast<uint8_t>(supportedModeMode);
+        option.semanticTags = DataModel::List<const SemanticTag>(semanticTags, semanticTagCount);
+
+        modeOptionStruct[index] = option;
+    }
+    endpointArray[endpointId][0] = modeOptionStruct;
+    endpointArray[endpointId][1] = (modeOptionStruct + supportedModeCount);
+
+    return ModeOptionsProvider(modeOptionStruct, modeOptionStruct + supportedModeCount);
+}
+
+Status StaticSupportedModesManager::getModeOptionByMode(unsigned short endpointId, unsigned char mode,
+                                                        const ModeOptionStructType ** dataPtr) const
+{
+    auto modeOptionsProvider = this->getModeOptionsProvider(endpointId);
+    if (modeOptionsProvider.begin() == nullptr)
+    {
+        return Status::UnsupportedCluster;
+    }
+    auto * begin = modeOptionsProvider.begin();
+    auto * end   = modeOptionsProvider.end();
+
+    for (auto * it = begin; it != end; ++it)
+    {
+        auto & modeOption = *it;
+        if (modeOption.mode == mode)
+        {
+            *dataPtr = &modeOption;
+            return Status::Success;
+        }
+    }
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "Cannot find the mode %u", mode);
+    return Status::InvalidCommand;
+}
+
+const ModeSelect::SupportedModesManager * ModeSelect::getSupportedModesManager()
+{
+    return &StaticSupportedModesManager::instance;
+}
+
+void StaticSupportedModesManager::FreeSupportedModes()
+{
+    for (int i = 0; i < FIXED_ENDPOINT_COUNT; i++)
+    {
+        if (endpointArray[i][0] != nullptr)
+        {
+            auto * begin = static_cast<ModeOptionStructType *>(endpointArray[i][0]);
+            auto * end   = static_cast<ModeOptionStructType *>(endpointArray[i][1]);
+            for (auto * it = begin; it != end; ++it)
+            {
+                auto & modeOption = *it;
+                delete[] modeOption.label.data();
+                delete[] modeOption.semanticTags.data();
```

Comment:
No, this still does not make any sense.  `modeOption.label.data()` is pointing somewhere into the _middle of a memory allocation_.

I would really like to understand how this is being tested.

Reply:
I tested this with similar sample code which I wrote and tested with `leaks` on mac and found zero memory leaks. 
[memoryLeak.cpp.zip](https://github.com/project-chip/connectedhomeip/files/11166729/memoryLeak.cpp.zip)

Attached the sample code. PTAL

Reply:
Please run with asan.  The default Mac allocator silently allows free() on pointers that were never allocated (as a no-op).... And the difference between your memoryLeak code and the code you are using here is that the memoryLeak code only runs once, not once per endpoint as you have here.

Reply:
@bzbarsky-apple The memory leaks has been tested with Espressif's [heap tracing](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/heap_debug.html#heap-tracing).

And got zero memory leaks